saffron/src/folding.rs

```diff
@@ -0,0 +1,695 @@
+    q: Evaluations<ScalarField, R2D<ScalarField>>,
+    a: Evaluations<ScalarField, R2D<ScalarField>>,
+}
+
+#[derive(PartialEq, Eq)]
+/// Relaxed witness extends the non-relaxed witness with evaluations
+/// of the error term.
+pub struct RelaxedWitness {
+    core: CoreWitness,
+    e: Evaluations<ScalarField, R2D<ScalarField>>,
+}
+
+impl CoreWitness {
+    pub fn relax(self, domain: R2D<ScalarField>) -> RelaxedWitness {
+        RelaxedWitness {
+            core: self,
+            e: Evaluations::from_vec_and_domain(vec![ScalarField::zero(); domain.size()], domain),
+        }
+    }
+}
+
+impl RelaxedInstance {
+    /// This function checks if the provided instance is valid regarding the
+    /// provided witness, by checking both the validity of the witness and its
+    /// consistency with the instance's commitments.
+    pub fn check_in_language(
+        &self,
+        srs: &SRS<Curve>,
+        domain: R2D<ScalarField>,
+        wit: &RelaxedWitness,
+    ) -> bool {
+        for i in 0..SRS_SIZE {
+            // todo can be parallelized
+            if self.u * wit.core.a[i] - wit.core.q[i] * wit.core.d[i] + wit.e[i]
+                != ScalarField::zero()
+            {
+                return false;
+            }
+        }
+        if self.core.comm_a
+            != srs
+                .commit_evaluations_non_hiding(domain, &wit.core.a)
+                .chunks[0]
+        {
+            return false;
+        }
+        if self.core.comm_d
+            != srs
+                .commit_evaluations_non_hiding(domain, &wit.core.d)
+                .chunks[0]
+        {
+            return false;
+        }
+        if self.core.comm_q
+            != srs
+                .commit_evaluations_non_hiding(domain, &wit.core.q)
+                .chunks[0]
+        {
+            return false;
+        }
+        if self.comm_e != srs.commit_evaluations_non_hiding(domain, &wit.e).chunks[0] {
+            return false;
+        }
+        true
+    }
+}
+
+// the first instance/witness is supposed to be non-relaxed
+pub fn folding_prover(
+    srs: &SRS<Curve>,
+    domain: R2D<ScalarField>,
+    inst1: &CoreInstance,
+    wit1: &CoreWitness,
+    inst2: &RelaxedInstance,
+    wit2: &RelaxedWitness,
+) -> (RelaxedInstance, RelaxedWitness, Curve) {
+    let mut fq_sponge = CurveFqSponge::new(Curve::other_curve_sponge_params());
+
+    let cross_term: Evaluations<ScalarField, R2D<ScalarField>> =
+        &(&(&wit2.core.a + &(&wit1.a * inst2.u)) - &(&wit2.core.q * &wit1.d))
+            - &(&wit1.q * &wit2.core.d);
+
+    let cross_term_comm: Curve = srs
+        .commit_evaluations_non_hiding(domain, &cross_term.clone())
+        .chunks[0];
+
+    fq_sponge.absorb_g(&[
+        inst1.comm_d,
+        inst1.comm_q,
+        inst1.comm_a,
+        inst2.core.comm_d,
+        inst2.core.comm_q,
+        inst2.core.comm_a,
+        inst2.comm_e,
+        cross_term_comm,
+    ]);
+
+    let recombination_chal = fq_sponge.squeeze(2);
```

Comment:
Nit: as mentioned in https://github.com/o1-labs/proof-systems/pull/3183, we can use `challenge`.

Reply:
I think `recombination_chal` is more descriptive: aside from it, we also have challenge seed for storage proof (`challenge_seed`), the challenge itself `challenge` (for recombining!), in other plonkish proofs we have evaluation challenge `zeta` (a point, I call it `evaluation_point`), and we have two challenges `polyscale` and `evalscale` for recombining polynomials. Everything verifier samples is a challenge, but I'm thinking if it's cheap to give it a distinct name, it makes it more clear?

Reply:
I meant
```
let recombination_chal = fq_sponge.challenge()
```